File: src/main/java/com/tambapps/http/restclient/response/ErrorResponse.java
package com.tambapps.http.restclient.response;

import com.tambapps.http.restclient.response.handler.ResponseHandler;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Collections;
import java.util.List;

public class ErrorResponse<T> extends AbstractRestResponse<T> {

  private final byte[] bytes;

  public ErrorResponse() { //no response
    this(REQUEST_NOT_SENT, new HttpHeaders(Collections.<String, List<String>>emptyMap()));
  }

  public ErrorResponse(int responseCode, HttpHeaders headers) {
    this(responseCode, headers, null);
  }

  public ErrorResponse(int responseCode, HttpHeaders headers, byte[] bytes) {
    super(responseCode, headers);
    this.bytes = bytes;
  }

  @Override
  public boolean isErrorResponse() {
    return true;
  }

  @Override
  public boolean isSuccessful() {
    return false;
  }

  @Override
  public T getData() {
    return null;
  }

  @Override
  public <ErrorT> ErrorT getErrorData(ResponseHandler<ErrorT> responseHandler) {
    if (bytes == null) {
      return null;
    }
    try (ByteArrayInputStream is = new ByteArrayInputStream(bytes)) {
      return responseHandler.convert(is);
    } catch (IOException e) {
      return null;
    }
  }

  @Override
  public byte[] getRawErrorData() {
    return bytes;
  }

}

File: src/main/kotlin/com/tambapps/http/client/response/ErrorResponse.kt
package com.tambapps.http.client.response

import com.tambapps.http.client.response.handler.ResponseHandler
import java.io.ByteArrayInputStream

class ErrorResponse<T> constructor(responseCode: Int, headers: HttpHeaders, override val rawErrorData: ByteArray) : AbstractResponse<T>(responseCode, headers) {

    constructor(rawErrorData: ByteArray): this(Response.REQUEST_NOT_SENT, HttpHeaders(emptyMap()), rawErrorData)

    override val isErrorResponse = true
    override val isSuccessful = false
    override val hasBody = rawErrorData.isNotEmpty()

    override val data: T get() = throw IllegalAccessException("Error response doesn't have error data")

    override fun <ErrorT> getErrorData(responseHandler: ResponseHandler<ErrorT>): ErrorT {
        return ByteArrayInputStream(rawErrorData).use { iStream ->  responseHandler.convert(iStream) }
    }

}